### PR TITLE
Third lab

### PR DESCRIPTION
**Pull request by ShevchenyaIlya**
Third laboratory work. Battleship game.

Выделение_247
Выделение_248
Выделение_249

### CODE REVIEW

File: Battleship/app/src/main/java/com/example/battleship/ProfileActivity.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+                        SimpleAdapter adapter = new SimpleAdapter(context, arrayList, android.R.layout.simple_list_item_2,
+                                new String[]{"email", "result"},
+                                new int[]{android.R.id.text1, android.R.id.text2});
+                        scores.setAdapter(adapter);
+                    }
+                });
+
+        if(userConnection.getUser().getImageUrl() != null)
+        {
+            StorageReference photoReference = storageReference.child(userConnection.getUser().getImageUrl());
```
Comment (gmltA) [inline Battleship/app/src/main/java/com/example/battleship/ProfileActivity.java:100]: Gravatar avatar source is not implemented. See requirements.

File: Battleship/app/src/main/java/com/example/battleship/ConnectionActivity.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        userConnection = new UserConnection((User)getIntent().getSerializableExtra("User"), "");
+    }
+
+    public void createConnect(View view) {
+        Map<String, Object> connection = new HashMap<>();
+        connection.put("sender", userConnection.getUser().getEmail());
+        connection.put("recipient", "");
+        final Context context = this;
+        db.collection("connections").add(connection);
+        db.collection("connections")
```
Comment (gmltA) [inline Battleship/app/src/main/java/com/example/battleship/ConnectionActivity.java:54]: Network interaction should be extracted from activities (UI layer) into separate independent services.

File: Battleship/app/src/main/java/com/example/battleship/Game.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import androidx.lifecycle.ViewModel;
+
+import com.google.firebase.firestore.FirebaseFirestore;
+
+public class Game extends ViewModel {
+    private BattleField yourField;
+    private BattleField opponentField;
+
+    private boolean horizontalOrientation = true;
+    private String selectedShip = "Four-decker";
```
Comment (gmltA) [inline Battleship/app/src/main/java/com/example/battleship/Game.java:12]: Consider moving vessel types to Enums (or even separate class).

File: Battleship/app/src/main/java/com/example/battleship/Game.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    private String mode = "Create";
+    private String documentId;
+
+    private UserConnection userConnection;
+    private String connectionId;
+    private String opponentEmail;
+    private String opponentId;
```
Comment (gmltA) [inline Battleship/app/src/main/java/com/example/battleship/Game.java:19]: Wouldn't it be better to have some kind of `Player` class with all this data?

File: Battleship/app/src/main/java/com/example/battleship/LoginActivity.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                                    Log.d(TAG, "Error getting documents: ", task.getException());
+                                }
+                            }
+                        });
+            }
+            else
+                Toast.makeText(this, "Short password or wrong email.", Toast.LENGTH_SHORT).show();
+        }
+        else
+            Toast.makeText(this, "Fill all field.", Toast.LENGTH_SHORT).show();
```
Comment (gmltA) [inline Battleship/app/src/main/java/com/example/battleship/LoginActivity.java:86]: All this auth-related stuff should be extracted in `AuthService` class.

File: Battleship/app/src/main/java/com/example/battleship/LoginActivity.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    }
+
+    @Override
+    protected void onSaveInstanceState(@NonNull Bundle outState) {
+        outState.putString("login", login.getText().toString());
+        outState.putString("password", password.getText().toString());
```
Comment (gmltA) [inline Battleship/app/src/main/java/com/example/battleship/LoginActivity.java:93]: What about ViewModels we learned previously?

File: Battleship/app/src/main/java/com/example/battleship/MainActivity.java [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+        score.put("winner", winner);
+        score.put("looser", looser);
+        db.collection("scores").add(score);
+    }
+
+    private void loadOpponentField(final String opponent)
+    {
+        game.setOpponentEmail(opponent);
+        final Context context = this;
+        db.collection("collections")
```
Comment (gmltA) [inline Battleship/app/src/main/java/com/example/battleship/MainActivity.java:203]: Such level of nested callbacks gives you a hint that something is not right in the architecture)

**Review by gmltA — CHANGES_REQUESTED** [review]
Good job. But some improvements would make it even better,

**Review by gmltA — APPROVED** [review]
That's better! Good luck during the exam!

**Comment by ShevchenyaIlya** [comment]
@gmltA Thank you! However I did not correct everything from what you indicated above. Or is it enough to pass this laboratory work?

**Comment by gmltA** [comment]
> @gmltA Thank you! However I did not correct everything from what you indicated above. Or is it enough to pass this laboratory work?

That's enough.

**Pull request merged**